JWST + HIRES

jhennawi opened the pull request:
This PR is meant to start building the tools for JWST NIRSPEC reductions and full echelle reductions with Keck/HIRES. Here is a summary of the changes and some comments: 

-- For complicated reasons related to conflicts, the HIRES and JWST branches have been merged. Apologies to the reviewers that you have to review this code twice. 
-- Most of the HIRES changes are to HIRES specific modules that are not used yet for full reductions, although some core wavelength calibration routines needed to be modified to accommodate some of the functionality that will be needed for HIRES wavelength calibration. This will impact ALL wavelength solutions, and the QA plots need to be inspected carefully before we merge. 
-- The JWST changes significantly modify the find_objects.py and extraction.py base classes. These changes provide the flexibility to use these classes in the absence of a full PypeIt calibration processing run, which is impossible with JWST. 
-- Coadd2d now operates on images of different sizes, which was a non-trivial but very important upgrade.

kbwestfall committed:
allow setup to keep bad configurations

kbwestfall committed:
Merge branch 'release' into setup_bad_frames

kbwestfall committed:
Merge branch 'develop' into setup_bad_frames

jhennawi committed:
New branch for HIRES dev

jhennawi committed:
Getting hires started!

Commit:
image reading

jhennawi committed:
pypeit par and HIRES integration.

jhennawi committed:
pypeit par and HIRES integration.

jhennawi committed:
pypeit par and HIRES integration.

Commit:
template file

jhennawi committed:
pypeit par and HIRES integration.

jhennawi committed:
pypeit par and HIRES integration.

Commit:
load

Commit:
Merge branch 'keck_hires_dev' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
pypeit par and HIRES integration.

Commit:
wavelengths

jhennawi committed:
pypeit par and HIRES integration.

jhennawi committed:
Fixing a minor bug

jhennawi committed:
Fixing a minor bug

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
Fixing a minor bug

jhennawi committed:
Improving for HIRES.

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst
	pypeit/core/arc.py

jhennawi committed:
Needed this for the plane. Back to normal.

kbwestfall committed:
Merge branch 'develop' into setup_bad_frames

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
tweaks to wavelength utils codes.

jhennawi committed:
Merge branch 'objfind_refactor' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
mosaic

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
Fixed bug with maxnumber logic.

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
Merge branch 'release' of https://github.com/pypeit/PypeIt into keck_hires_dev

kbwestfall committed:
keck hires mosaic trace

kbwestfall committed:
bug

jhennawi committed:
combspec now returns the outmask

jhennawi committed:
Merge branch 'hires_mosaic' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
combspec now returns the outmask

jhennawi committed:
New module holding echelle routines.

jhennawi committed:
HIRES wavelengths.

jhennawi committed:
HIRES wavelengths.

jhennawi committed:
ra sorted the list

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	pypeit/core/wavecal/autoid.py
	pypeit/wavecalib.py

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
Fixed bug when nivar_opt was equal to none.

kbwestfall committed:
Merge branch 'release' into keck_hires_dev

kbwestfall committed:
basic doc fixes

kbwestfall committed:
inline comments

kbwestfall committed:
Merge branch 'develop' into keck_hires_dev

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
Merge branch 'keck_hires_dev' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
Minor

jhennawi committed:
Added channel label to QA.

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

 Conflicts:
	CHANGES.rst

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into jwst

jhennawi committed:
Added channel label to QA.

jhennawi committed:
Added channel label to QA.

jhennawi committed:
Removing calibrate dependency.

jhennawi committed:
Refactoring base clases dependencies.

jhennawi committed:
Getting new base classes working.

jhennawi committed:
Getting new base classes working.

jhennawi committed:
Getting new base classes working.

jhennawi committed:
Some minor tweaks to coadd2d and skysubtraction.

jhennawi committed:
Coadd2d and coadd core routines now works on lists of images instead of image stack arrays.

jhennawi committed:
Coadd2d and coadd core routines now works on lists of images instead of image stack arrays.

jhennawi committed:
Coadd2d and coadd core routines now works on lists of images instead of image stack arrays.

Commit:
Merge branch 'develop' into jwst_rjc

Commit:
Merge branch 'jwst' into jwst_rjc

Commit:
update find objects

jhennawi committed:
Merge branch 'keck_hires_dev' of https://github.com/pypeit/PypeIt into jwst_coadd2d_refactor

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into jwst_coadd2d_refactor

 Conflicts:
	pypeit/extraction.py

jhennawi committed:
Fixed bug associated with tilts in 2d coadding.

jhennawi committed:
Fixed bug associated with tilts in 2d coadds.

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
Addressing PR comments.

jhennawi committed:
Trying to fix a bug in SINFONI.

jhennawi committed:
Make cross-correlations more robust for wavelengths and other echelle modifications.

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into keck_hires_dev

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into jwst_coadd2d_refactor

 Conflicts:
	CHANGES.rst
	doc/include/shane_kast_blue_A.pypeit.rst
	doc/pypeit_par.rst
	pypeit/core/coadd.py
	pypeit/core/combine.py
	pypeit/core/skysub.py
	pypeit/par/pypeitpar.py

jhennawi committed:
Make cross-correlations more robust for wavelengths and other echelle modifications.

jhennawi committed:
Merge branch 'keck_hires_dev' of https://github.com/pypeit/PypeIt into jwst_coadd2d_refactor

 Conflicts:
	pypeit/core/coadd.py

jhennawi committed:
Make cross-correlations more robust for wavelengths and other echelle modifications.

jhennawi committed:
Merge branch 'keck_hires_dev' of https://github.com/pypeit/PypeIt into jwst_coadd2d_refactor

 Conflicts:
	CHANGES.rst
	doc/pypeit_par.rst
	pypeit/core/coadd.py
	pypeit/core/skysub.py
	pypeit/par/pypeitpar.py

jhennawi committed:
Adding the option to not skysub in object finding.

jhennawi committed:
Removed the options to not skysub.

jhennawi committed:
Removed the options to not skysub.

jhennawi committed:
Added back options to skip skysub.

jhennawi committed:
Added back options to skip skysub.

jhennawi committed:
Something got messed up merging I think. Fixing this parset.

jhennawi committed:
Added waveimg_mid_pseudo and patching gaps in the waveimg.

jhennawi committed:
Global skysub option removed since it is now part of find_objects.

jhennawi committed:
Minor parset change

jhennawi committed:
Test.

jhennawi committed:
Test.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -204,10 +204,10 @@ def __init__(self, spec2d, spectrograph, par, det=1, offsets=None, weights='auto
-            msgs.error('Not all of your exposures have the same spectral dimension. Check your inputs')
-        self.nspec = self.stack_dict['slits_list'][0].nspec
+        self.nspec_array = np.array([slits.nspec for slits in self.stack_dict['slits_list']])
+        #if not len(set(nspec_list)) == 1:
```
remove?

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -364,8 +364,10 @@ def coadd(self, only_slits=None, interp_dspat=True):
                                                          objid=self.objid_bri)
-            thismask_stack = np.abs(self.stack_dict['slitmask_stack'] - self.stack_dict['slits_list'][0].spat_id[slit_idx]) <= self.par['coadd2d']['spat_toler']
 
+            #thismask_stack = np.abs(self.stack_dict['slitmask_stack'] - self.stack_dict['slits_list'][0].spat_id[slit_idx]) <= self.par['coadd2d']['spat_toler']
```
remove?

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -619,22 +625,23 @@ def reduce(self, pseudo_dict, show=None, show_peaks=None, basename=None):
-        exTract.binning = self.binning
-        exTract.basename = basename
-        exTract.reduce_bpm = pseudo_reduce_bpm
+        #exTract.tilts = pseudo_dict['tilts']
```
remove?

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -709,10 +716,11 @@ def offset_slit_cen(self, slitid, offsets):
         """
-        ref_trace_stack = np.zeros((self.stack_dict['slits_list'][0].nspec, len(offsets)),
-                                   dtype=float)
+        #ref_trace_stack = np.zeros((self.stack_dict['slits_list'][0].nspec, len(offsets)),
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -828,39 +847,46 @@ def load_coadd2d_stacks(self, spec2d, chk_version=False):
+            slitmask_stack.append(s2dobj.slits.slit_img(flexure=s2dobj.sci_spat_flexure))
+            #tilts_stack.append(s2dobj.tilts)
+
+        #    if ifile == 0:
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -1081,27 +1107,28 @@ def compute_offsets(self, offsets):
-            var_list = [utils.inverse(self.stack_dict['sciivar_stack'])]
+            sci_list = [[sciimg - skymodel for sciimg, skymodel in zip(self.stack_dict['sciimg_stack'], self.stack_dict['skymodel_stack'])]]
+            var_list = [[utils.inverse(sciivar) for sciivar in self.stack_dict['sciivar_stack']]]
+            #sci_list = [self.stack_dict['sciimg_stack'] - self.stack_dict['skymodel_stack']]
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -1181,7 +1208,7 @@ def get_brightest_obj(self, specobjs_list, spat_ids):
         msgs.info('Finding brightest object')
         nexp = len(specobjs_list)
-        nspec = specobjs_list[0][0].TRACE_SPAT.shape[0]
+        #nspec = specobjs_list[0][0].TRACE_SPAT.shape[0]
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -1294,21 +1322,38 @@ def get_maskdef_dict(self, slit_idx, ref_trace_stack):
+                slit_cen_dspat_vec[iexp] = (maskdef_slitcen_pixpos - ref_trace[nspec_this//2])/self.spat_samp_fact
+
+            imaskdef_slitcen_dspat = np.mean(slit_cen_dspat_vec)
+            #imaskdef_slitcen_dspat = np.mean((maskdef_slitcen_pixpos - ref_trace_stack[self.nspec//2, :])/self.spat_samp_fact)
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/coadd2d.py:
```diff
@@ -1294,21 +1322,38 @@ def get_maskdef_dict(self, slit_idx, ref_trace_stack):
+                objpos_dspat_vec[iexp] = (maskdef_obj_pixpos - ref_trace[nspec_this//2])/self.spat_samp_fact
+
+            imaskdef_objpos_dspat = np.mean(objpos_dspat_vec)
+            #imaskdef_objpos_dspat = np.mean((maskdef_obj_pixpos - ref_trace_stack[self.nspec//2, :])/self.spat_samp_fact)
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/core/coadd.py:
```diff
@@ -91,7 +91,10 @@ def renormalize_errors(chi, mask, clip = 6.0, max_corr = 5.0, title = '', debug=
         mask (`numpy.ndarray`_):
             True = good, mask for your chi array of type bool
-        clip (float, optional): 
+        clip (float, optional):
```
add a description

jhennawi replied:
Done.

profxj commented on pypeit/core/coadd.py:
```diff
@@ -91,7 +91,10 @@ def renormalize_errors(chi, mask, clip = 6.0, max_corr = 5.0, title = '', debug=
+        clip (float, optional):
+        mask (`numpy.ndarray`_):
+            True = good, mask for your chi array of type bool
+        clip (float, optional):
```
does this appear twice?

jhennawi replied:
Fixed.

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2082,24 +2086,25 @@ def combspec(waves, fluxes, ivars, masks, sn_smooth_npix,
         percentile fraction cut used for selecting minimum SNR cut for robust_median_ratio
     maxiter_scale: int, optional, default=5
         Maximum number of iterations performed for rescaling spectra.
+    wave_grid_input (`numpy.ndarray`_):
```
I think this is the wrong doc style which will break on readthedocs

jhennawi replied:
I need help here.

kbwestfall replied:
Change to 
```
wave_grid_input : `numpy.ndarray`_
```

kbwestfall replied:
fixed

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2692,16 +2705,15 @@ def get_wave_bins(thismask_stack, waveimg_stack, wave_grid):
-
-    wave_grid : `numpy.ndarray`_
-        shape (ngrid)
+    thismask_stack (list):
```
wrong doc style.
should be:

`thismask_stack: list`

same for most of the rest of this doc string

jhennawi replied:
fixed

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2731,14 +2749,22 @@ def get_spat_bins(thismask_stack, trace_stack, spat_samp_fact=1.0):
-        Array holding the stack of traces for each image in the stack. This is either the trace of the center of the slit
-        or the trace of the object in question that we are stacking about.
-    spat_samp_fact: float, optional
+    thismask_stack (list):
```
same problem as noted above

jhennawi replied:
Fixed.

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2810,58 +2830,68 @@ def compute_coadd2d(ref_trace_stack, sciimg_stack, sciivar_stack, skymodel_stack
-            changed here, but am leaving it for now.)
-        interp_dspat (bool, optional):
-            Interpolate the dspat images wherever the coadds are masked
+        interp_dspat : bool, optional
```
wrong doc style

jhennawi replied:
I need assistance with doc string formatting.

kbwestfall replied:
`interp_dspat (bool, optional):`

kbwestfall replied:
fixed

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2810,58 +2830,68 @@ def compute_coadd2d(ref_trace_stack, sciimg_stack, sciivar_stack, skymodel_stack
 
 
     Returns:
         dict: Returns a dict with the following keys:
```
I think `Returns a dict` needs to be on the next line.

jhennawi replied:
I need help with doc string formatting.

kbwestfall replied:
This works fine.

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2925,7 +2959,8 @@ def compute_coadd2d(ref_trace_stack, sciimg_stack, sciivar_stack, skymodel_stack
                                sigma_clip_stack=sci_list_rebin[2], sigrej=sigrej,
                                maxiters=maxiters)
-    sciimg, imgminsky, tilts, waveimg, dspat = sci_list_out
+    #sciimg, imgminsky, tilts, waveimg, dspat = sci_list_out
```
remove?

jhennawi replied:
Done.

profxj commented on pypeit/core/coadd.py:
```diff
@@ -2977,29 +3014,23 @@ def compute_coadd2d(ref_trace_stack, sciimg_stack, sciivar_stack, skymodel_stack
+                dspat=dspat, nspec=imgminsky.shape[0], nspat=imgminsky.shape[1],
+                maskdef_id=maskdef_id, maskdef_slitcen=new_maskdef_slitcen, maskdef_objpos=new_maskdef_objpos,
                 maskdef_designtab=maskdef_designtab)
+#                rebin_weights_stack=sci_list_rebin[0], rebin_sciimg_stack=sci_list_rebin[1],
```
remove?

jhennawi replied:
Leave for now may be needed for debugging.

profxj commented on pypeit/core/combine.py:
```diff
@@ -261,3 +257,48 @@ def broadcast_weights(weights, shape):
+    Utility routine to broadcast weights to be the size of image stacks specified by shape
+
+    Args:
+        weights: :obj:`list`
```
wrong doc style.  Either change to `Parameters` or use Google style.

profxj commented on pypeit/core/combine.py:
```diff
@@ -261,3 +257,48 @@ def broadcast_weights(weights, shape):
+
+
+    Returns:
+        List of `numpy.ndarray` weight images where each image in the list has shape that was input via the shapes
```
needs to be a datatype

profxj commented on pypeit/core/skysub.py:
```diff
@@ -97,6 +95,9 @@ def global_skysub(image, ivar, tilts, thismask, slit_left, slit_righ,
             requires that the counts are positive which will not be the
             case for i.e. an A-B image. Thus the routine will fail if
             pos_mask is not set to False.
+        max_mask_frac: float, default max_bad_pixel_frac = 0.80
```
wrong doc style

kbwestfall replied:
fixed

profxj commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,213 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+    spec :  float ndarray shape of (nspec, norders)
+        Array of arc spectra for each order for which wavelength solutions are
+        desired.
+    orders : `numpy.ndarray`, optional
```
does not appear `optional`

profxj replied:
and missing the `_`

jhennawi replied:
Fixed.

profxj commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,213 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+
+    Parameters
+    ----------
+    spec :  float ndarray shape of (nspec, norders)
```
Needs to be `numpy.ndarray`_

jhennawi replied:
Fixed.

profxj commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,213 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+        Order numbers for the provided spectra. Used to match against
+        the relevant archived spectrum for echelle spectrographs.
+        Shape must be :math:`(N_{\rm orders},)`
+    spec_arxiv :  float ndarray shape (nspec, narxiv) or (nspec)
```
needs to be a type, e.g. numpy.ndarray

jhennawi replied:
Fixed.

profxj commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,213 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+        threshold are not used in wavelength solution fits because
+        they cannot be accurately centroided
+
+
```
Is nothing Return'ed?

jhennawi replied:
Added.

profxj commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1138,14 +1077,15 @@ class ArchiveReid:
     par : :class:`~pypeit.par.pypeitpar.WaveSolutionPar`
         Key parameters that drive the behavior of the
         wavelength-solution algorithms.
+    ech_fixed_format (bool):
```
wrong doc style

jhennawi replied:
Need help with the docs.

profxj commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1227,7 +1167,7 @@ def __init__(self, spec, spectrograph, lamps, par, ok_mask=None, measured_fwhms=
 
         self.par = par
-        self.spectrograph = spectrograph
+        #self.spectrograph = spectrograph
```
remove

jhennawi replied:
Done.

profxj commented on pypeit/core/wavecal/echelle.py:
```diff
@@ -0,0 +1,217 @@
+            Number of orders to pad the coverage by on the blue and red side.
+
+    Returns:
+        order_vec (numpy.ndarray):
```
can't be a variable name

jhennawi replied:
Need help with the doc formatting.

kbwestfall replied:
Mimic what's done below in `identify_ech_orders`.

profxj commented on pypeit/core/wavecal/echelle.py:
```diff
@@ -0,0 +1,217 @@
+            Number of spectral pixels in the echelle spectrum
+
+    Returns:
+        wave_soln_guess (numpy.ndarray):
```
same as above

jhennawi replied:
Need help with doc formatting. I like variable names in return docs.

kbwestfall replied:
Mimic what's done below in `identify_ech_orders`.

kbwestfall replied:
fixed

profxj commented on pypeit/core/wavecal/echelle.py:
```diff
@@ -0,0 +1,217 @@
+        pad (int):
+            Number of orders to pad the coverage by on the blue and red side.
+    Returns:
+
```
What is returned?  And you likely need a blank line above `Returns:`

jhennawi replied:
Added.

kbwestfall replied:
wrong format. Mimic what's done below in `identify_ech_orders`.

kbwestfall replied:
fixed

profxj commented on pypeit/core/wavecal/templates.py:
```diff
@@ -430,6 +437,64 @@ def xidl_arcspec(xidl_file, slit):
+            Keck/HIRES save file
+
+    Returns:
+
```
fill in

jhennawi replied:
Actually you wrote this method, so please help out!

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -199,81 +204,82 @@ def get_wave_grid(waves, masks=None, wave_method='linear', iref=0, wave_grid_min
+
+        dwave_data, dloglam_data, resln_guess, pix_per_sigma = get_sampling(waves)
+
+        # TODO: These tests of the string value should not use 'in', they should use ==
```
why?  `in` works the same and is nicer to read.

jhennawi replied:
TODO removed.

kbwestfall replied:
I added that todo.  Previously (look in `develop`) this code has lines like, `elif 'linear' in wave_method`.  We should be testing equality in this case, not whether or not the string is a subset of the other string.  To my mind it's bad coding practice.  Here's a contrived, but somewhat relevant example:
```
>>> wave_method = 'log-linear'
>>> 'linear' in wave_method
True
>>> wave_method == 'linear'
False
```

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -390,49 +396,90 @@ def zerolag_shift_stretch(theta, y1, y2):
+    the input spectrum.
+
+    Args:
+        inspec1 (ndarray):
```
numpy.ndarray_

kbwestfall replied:
fixed.

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -390,49 +396,90 @@ def zerolag_shift_stretch(theta, y1, y2):
 
-    # Run line detection to get the continuum subtracted arc
+    Returns:
+        xcorr_arc (ndarray):
```
can't use the variable name.
and should be numpy.ndarray_

jhennawi replied:
I like the variable name and this is the sklearn style which I've been pushing for.

jhennawi replied:
Need help with docs.

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -390,49 +396,90 @@ def zerolag_shift_stretch(theta, y1, y2):
+        else:
+            xcorr_arc += ampl_clip[ind]*np.exp(-0.5*((spec_vec - tcent1[ind])/sigma)**2)
+
+    #for ind in w1:
```
rmeove?

jhennawi replied:
Done.

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -444,9 +491,12 @@ def xcorr_shift(inspec1,inspec2, smooth=1.0, percent_ceil=80.0, use_raw_arc=Fals
-        smooth: float, default=1.0
-            Gaussian smoothing in pixels applied to both spectra for the
-            computations. Default is 5.0
+        sigdetect (float, optional, default=3.0):
```
wrong doc style (or at least different from what is used for `inspec2`)

jhennawi replied:
Need help with docs.

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -521,9 +579,13 @@ def xcorr_shift_stretch(inspec1, inspec2, cc_thresh=-1.0, smooth=1.0, percent_ce
-    smooth: float, default
-        Gaussian smoothing in pixels applied to both spectra for the computations. Default is 5.0
-    percent_ceil: float, default=90.0
+    sigdetect (float, optional, default=3.0):
```
wrong doc style

jhennawi replied:
Need help with docs.

kbwestfall replied:
`sigdetect : float, optional, default=3.0`

kbwestfall replied:
fixed

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -655,21 +724,34 @@ def wavegrid(wave_min, wave_max, dwave, spec_samp_fact=1.0, log10=False):
     Returns:
-        `numpy.ndarray`_: Wavelength grid in Angstroms (i.e. log10 even
-        if log10 is requested)
+        :obj:`tuple`: Returns two `numpy.ndarray`_ objects and a float:
```
Description needs to be on a new line, I think..

jhennawi replied:
fixed.

profxj commented on pypeit/extraction.py:
```diff
@@ -126,7 +139,7 @@ def __init__(self, sciImg, sobjs_obj, spectrograph, par, caliBrate,
         self.objtype = objtype
         self.par = par
-        self.caliBrate = caliBrate
+        #self.caliBrate = caliBrate
```
fine.
remove

jhennawi replied:
Done.

profxj commented on pypeit/extraction.py:
```diff
@@ -181,6 +194,40 @@ def __init__(self, sciImg, sobjs_obj, spectrograph, par, caliBrate,
+        if waveTilts is None and tilts is None:
+            msgs.error("Must provide either waveTilts or tilts to FindObjects")
+        elif waveTilts is not None and tilts is not None:
+            msgs.error("Cannot provide both waveTilts and tilts to FindObjects")
```
Is this `FindObjects` ?

jhennawi replied:
Fixed.

profxj commented on pypeit/extraction.py:
```diff
@@ -213,7 +260,7 @@ def nobj_to_extract(self):
         return None
 
-    def initialise_slits(self, initial=False):
+    def initialise_slits(self, slits, initial=False):
```
add `slits` to doc string

jhennawi replied:
Added.

profxj commented on pypeit/find_objects.py:
```diff
@@ -126,25 +130,26 @@ def get_instance(cls, sciImg, spectrograph, par, caliBrate, objtype, bkg_redux=F
 
-    def __init__(self, sciImg, spectrograph, par, caliBrate,
-                 objtype, bkg_redux=False, find_negative=False, std_redux=False, show=False,
+    def __init__(self, sciImg, slits, spectrograph, par, objtype, waveTilts=None, tilts=None, sky_region_file=None,
```
Is `slits` documented somewhere?

profxj commented on pypeit/find_objects.py:
```diff
@@ -207,9 +212,31 @@ def __init__(self, sciImg, spectrograph, par, caliBrate,
         self.slitshift = np.zeros(self.slits.nslits)  # Global spectral flexure slit shifts (in pixels) that are applied to all slits.
         self.vel_corr = None
 
+        # Deal with dynamically generated calibrations, i.e. the tilts. If the tilts are not input generate
```
This looks like duplicated code from `Extract` (or the other way).

Can we avoid?

profxj commented on pypeit/find_objects.py:
```diff
@@ -251,7 +278,7 @@ def create_skymask(self, sobjs_obj):
         return skymask
 
-    def initialise_slits(self, initial=False):
+    def initialise_slits(self, slits, initial=False):
```
add `slits` to doc string

jhennawi replied:
Done.

profxj commented on pypeit/find_objects.py:
```diff
@@ -277,7 +304,7 @@ def initialise_slits(self, initial=False):
 #        self.spatial_coo = self.slits.spatial_coordinates(initial=initial, flexure=self.spat_flexure_shift)
 
-    def run(self, std_trace=None, show_peaks=False):
+    def run(self, std_trace=None, show_peaks=False, show_skysub_fit=False):
```
update doc string

jhennawi replied:
Updated.

profxj commented on pypeit/find_objects.py:
```diff
@@ -587,21 +616,24 @@ def load_skyregions(self, skymask_init):
             # Setup the master frame name
-            master_dir = self.caliBrate.master_dir
-            master_key = self.caliBrate.fitstbl.master_key(0, det=self.det) + "_" + sciName
+            #master_dir = self.caliBrate.master_dir
```
@rcooke-ast -- will this break anything?

rcooke-ast replied:
Nope - this is all good. `sky_region_file` is passed in as an argument now.

profxj commented on pypeit/images/pypeitimage.py:
```diff
@@ -168,6 +170,12 @@ def from_pypeitimage(cls, pypeitImage):
         # Return
         return slf
 
+    # JFH This API or the documentation should make it clear what is actually required to perform operations with these
```
Isn't it provided in the `datamodel` ?

jhennawi replied:
I think you should make the init for the class different from the datamodels and make the parameters required to create an image required parameters and the rest should be optional. That is a clearer API. I get that the datamodels needs all to be optional, but you can always make the object class separate from the datamodel class.

kbwestfall replied:
I agree we should make this more clear.  The instantiation of the `DataContainer` does not require all of the init arguments to be optional.  See, e.g., `EdgeTraceSet`.

profxj commented on pypeit/par/pypeitpar.py:
```diff
@@ -4663,7 +4677,7 @@ def valid_telescopes():
         """
         return [ 'GEMINI-N','GEMINI-S', 'KECK', 'SHANE', 'WHT', 'APF', 'TNG', 'VLT', 'MAGELLAN', 'LBT', 'MMT', 
-                'KPNO', 'NOT', 'P200', 'BOK', 'GTC', 'SOAR', 'NTT', 'LDT']
+                'KPNO', 'NOT', 'P200', 'BOK', 'GTC', 'SOAR', 'NTT', 'LDT', 'JWST']
```
Nice!!

profxj replied:
I can't wait to see its longitude,latitutde and altitude!

profxj commented on pypeit/pypeit.py:
```diff
@@ -849,21 +854,18 @@ def extract_one(self, frames, det, sciImg, objFind, initial_sky, sobjs_obj):
         else:
             # Although exrtaction is not performed, still need to prepare some masks and the tilts
-            self.exTract.prepare_extraction()
+            #self.exTract.prepare_extraction()
```
remove

profxj commented on pypeit/scripts/coadd_2dspec.py:
```diff
@@ -191,6 +193,7 @@ def main(args):
             os.makedirs(master_dir)
 
         # Instantiate the sci_dict
+        # TODO Why do we need this sci_dict at all?? JFH
```
I think for @marlageha

profxj commented on pypeit/spectrographs/vlt_fors.py:
```diff
@@ -68,8 +68,8 @@ def default_pypeit_par(cls):
-        par['sensfunc']['polyorder'] = 8
-        par['sensfunc']['IR']['telgridfile'] = 'TelFit_Paranal_NIR_9800_25000_R25000.fits'
+        par['sensfunc']['polyorder'] = 5
+        par['sensfunc']['IR']['telgridfile'] = 'TelFit_Paranal_VIS_4900_11100_R25000.fits'
```
I thought I already fixed this..

jhennawi replied:
I think the merge had some problems honestly. Some spooky things like this happened here and there.

jhennawi replied:
Fixed.

profxj commented on pypeit/telescopes.py:
```diff
@@ -193,3 +193,12 @@ def __init__(self):
+        super(JWSTTelescopePar, self).__init__(name='JWST',
+                                              longitude=0.0,
+                                              latitude=0.0,
+                                              elevation=0.0,
```
you gotta give the `elevation`! :)

profxj reviewed (changes requested):
Tremendous effort here.

Mainly doc massaging so that things render 
properly on `readthedocs`

jhennawi committed:
Test.

jhennawi committed:
Added NIRCAM!

jhennawi committed:
Added NIRCAM!

debora-pe commented on pypeit/coadd2d.py:
```diff
@@ -1294,21 +1322,38 @@ def get_maskdef_dict(self, slit_idx, ref_trace_stack):
             # But here we take only the value at the mid point
-            maskdef_slitcen_pixpos = self.stack_dict['slits_list'][0].maskdef_slitcen[self.nspec//2, slit_idx] + self.maskdef_offset
+
+            # TODO JFH now that we allow different image sizes with lists is this correct?
```
@jhennawi I don't think this will work for images with different sizes, since `self.stack_dict['slits_list'].maskdef_slitcen` and `self.stack_dict['slits_list'].maskdef_objpos` depends on the slit tracing performed in each image (which will have different `MasterSlits`). I think `maskdef_slitcen_pixpos` and `imaskdef_objpos` will have to be lists/arrays and  `imaskdef_slitcen_dspat` and `imaskdef_objpos_dspat` cannot be the mean over every `iexp`, but they will have to be lists/arrays with one value for each image. 
I'll think how we can fix this.
However, I believe if coadd2d is run for images with the same `MasterSlits` it should still work the same.

debora-pe replied:
@jhennawi I tried running `coadd2d` for images with the **same size** and these changes do not work either. The script crashes. We need to fix it.

jhennawi replied:
I fixed these issues. I agree that different sizes requires some thought though. For now, things seem to work on the tests. Please also have a look.

Commit:
Merge branch 'jwst' into jwst_rjc

Commit:
doc updates

Commit:
doc update

Commit:
rm spatillum

Commit:
fix skyregions

Commit:
work

Commit:
multiple fixes

jhennawi committed:
Changed supported to False.

jhennawi committed:
Removed MMT references.

jhennawi committed:
Merge pull request #1478 from pypeit/jwst_nircam_wfss

Geting JWST/NIRCAM Wide-Field Slitless Spectroscopy Started

Commit:
wv_calib

Commit:
wv_calib

Commit:
wv fix

jhennawi committed:
Merge pull request #1480 from pypeit/jwst_rjc

JWST sky regions updates

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into jwst

 Conflicts:
	pypeit/coadd2d.py
	pypeit/extraction.py
	pypeit/find_objects.py
	pypeit/pypeit.py
	pypeit/scripts/ql_multislit.py

jhennawi committed:
Merged in develop.

jhennawi committed:
Merged in develop.

Commit:
small fixes

Commit:
rm todo

jhennawi committed:
Merged in develop.

Commit:
fix skyreg

kbwestfall committed:
fix some par

kbwestfall committed:
par + doc fixes

Commit:
improved skymask control flow

jhennawi committed:
Merge branch 'jwst_update' of https://github.com/pypeit/PypeIt into jwst

 Conflicts:
	pypeit/find_objects.py

jhennawi committed:
Merged in changes from Ryan and made some other ones

Commit:
simplify

Commit:
rm load_mask

jhennawi committed:
Merge pull request #1487 from pypeit/jwst_docupdate

JWST docs update + sky regions

kbwestfall committed:
Merge branch 'jwst' into jwst_par

kbwestfall committed:
doc fix

jhennawi committed:
Merge pull request #1486 from pypeit/jwst_par

Missing parameters

Commit:
minor doc updates

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into jwst

jhennawi committed:
Removing load_mask to clean up skymask fixes.

jhennawi committed:
Removing load_mask to clean up skymask fixes.

jhennawi committed:
Fixed a bug in coadd2d maskdef

jhennawi committed:
Fixed a bug in coadd2d maskdef

jhennawi committed:
Bug fixes and improvements in 2d coadding.

jhennawi committed:
Cleaning up for PR.

jhennawi committed:
Cleaning up for PR.

jhennawi committed:
Cleaning up for PR.

jhennawi committed:
Cleaning up for PR.

jhennawi committed:
Cleaning up for PR.

jhennawi committed:
Cleaning up for PR.

jhennawi committed:
Fixed a bug in echelle 2d coadding. Updated some docs.

kbwestfall commented on pypeit/calibrations.py:
```diff
@@ -763,9 +763,12 @@ def get_wv_calib(self):
             # Instantiate
             self.waveCalib = wavecalib.BuildWaveCalib(self.msarc, self.slits, self.spectrograph,
                                                       self.par['wavelengths'], lamps,
+                                                      meta_dict = meta_dict,
```
Comments from #1454 on this:

> @kbwestfall : Digging into wavecalib a bit, I think you're doing this so that you have access to echangle, xdangle, and dispname. Can this go into config_specific_par instead?

> @jhennawi : This is a good point @kbwestfall. I'm not sure what the right things is to do here, but could use your help and suggestions. We've used config_specific_par to set configuration specific parameters, but typically not use the parsets to hold observation specific information like grating, decker, etc. Echelle calibration requires these angles, and those are stored in the fitstable which is why I opted for the approach here. I'm not sure what the cleanest way to to code this is though. We need a rule for what goes in a parset and where to keep other more critical meta information.

Yeah, I see the difference, and agree this is a different take on config_specific_par than we have been using.  My main complaint is that I dislike passing around broadly or poorly defined objects when we only need a few items from them.  I'm not sure of the best solution, but it would be useful to talk it through.  Since this is HIRES specific, can you add a relevant "TODO" and make sure we come back to it for a future HIRES PR?

kbwestfall replied:
Added a TODO.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -451,17 +451,29 @@ def create_pseudo_image(self, coadd_list):
-            waveimg_pseudo[ispec, ispat] = coadd_dict['waveimg']
-            tilts_pseudo[ispec, ispat] = coadd_dict['tilts']
+            this_waveimg = coadd_dict['waveimg']
+            waveimg_pseudo[ispec, ispat] = this_waveimg
```
Do you need `this_waveimg`?  Why not `waveimg_pseudo[ispec, ispat] = coadd_dict['waveimg']`?

kbwestfall replied:
Removed.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -451,17 +451,29 @@ def create_pseudo_image(self, coadd_list):
             # spat_img_pseudo is the sub-pixel image position on the rebinned pseudo image
             inmask_pseudo[ispec, ispat] = coadd_dict['outmask']
-            image_temp = (coadd_dict['dspat'] - coadd_dict['dspat_mid'][0] + spat_left)*coadd_dict['outmask']
+            image_temp = (coadd_dict['dspat'] - coadd_dict['dspat_mid'][0] + spat_left) #*coadd_dict['outmask']
```
Just checking: is it okay that you've lost the masking here?

jhennawi replied:
Yeah that was done on purpose.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -550,6 +562,8 @@ def reduce(self, pseudo_dict, show=None, show_peaks=None, basename=None):
         show = self.show if show is None else show
         show_peaks = self.show_peaks if show_peaks is None else show_peaks
+        # TODO I think this is incorrect. I don't know how to instanatiate this class from an existing mask
+        # because everthying is hidden behind exceedingly complicated bitmasks.
```
Can you specify which part you don't think is correct?  FWIW, the bpm is just 1 for a bad pixel, 0 for a good pixel.  The `fullmask` attribute is the one that holds the "exceedingly complicated bitmask".

jhennawi replied:
The API for PypeIt image is impossible to understand because all the arguments are optional. Furthermore, sometimes we need to instantiate a PypeIt image from existing images without running the PypeIt image processing. In those situations (i.e. JWST, here in coadd2d, but also in quicklook), one has a mask for the image already. This is the combination of a bad pixel mask, CR mask, and possibly other masking. In this situation the mask does not adhere to our bit systems which is anyway. In those instances I need to know how to instantiate a PypeIt image with an existing mask. Improvements would entail making the API have required arguments, since this everything is optional is a very poor model. And also possibly more clear information documentation on how to input a mask. You might think that the crmask is the thing to use here but I already went that route and it is incorrect since the CRmask is constantly updated when you run LAcosmic in the base classes, whereas I want a mask that is basically always there. Can you please explain how to instantiate the fullmask then from an existing boolean good pixel mask? I suspect this is wrong in multiple places.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -699,19 +696,22 @@ def offsets_report(self, offsets, offsets_method):
 
         Returns:
+            ref_trace_stack (list):
+               A list of reference traces for the 2d coadding that have been offset
```
Wrong doc format.  If you want to give the name of the returned object, you have to use numpy format.  Copy the style and syntax in the `optimal_weights` doc string.

kbwestfall replied:
Fixed syntax.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -699,19 +696,22 @@ def offsets_report(self, offsets, offsets_method):
         for iexp, slits in enumerate(self.stack_dict['slits_list']):
-            ref_trace_stack[:,iexp] = slits.center[:,slitid] - offsets[iexp]
+            ref_trace_stack.append(slits.center[:,slitid] - offsets[iexp])
         return ref_trace_stack
```
I'm not sure, but it may be more efficient to use list compression:
```
ref_trace_stack = [slits.center[:,slitid] - offsets[iexp] for iexp, slits in enumerate(self.stack_dict['slits_list'])]
```

kbwestfall replied:
Changed.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -1080,27 +1077,26 @@ def compute_offsets(self, offsets):
-            #                                       self.stack_dict['tslits_dict_list'][iexp]['slit_righ'][:,slitid_bri])/2.0
+            trace_stack_bri = []
+            for slits in self.stack_dict['slits_list']:
+                trace_stack_bri.append(slits.center[:, self.slitidx_bri])
```
Use list compression?

kbwestfall replied:
Changed.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -1080,27 +1077,26 @@ def compute_offsets(self, offsets):
-            sci_list = [self.stack_dict['sciimg_stack'] - self.stack_dict['skymodel_stack']]
-            var_list = [utils.inverse(self.stack_dict['sciivar_stack'])]
+            sci_list = [[sciimg - skymodel for sciimg, skymodel in zip(self.stack_dict['sciimg_stack'], self.stack_dict['skymodel_stack'])]]
+            var_list = [[utils.inverse(sciivar) for sciivar in self.stack_dict['sciivar_stack']]]
```
Can we use numpy arrays to do this, or does the shape of the images vary?

jhennawi replied:
The whole point of refactoring everything to use lists was to accommodate coadding images of different sizes.

kbwestfall commented on pypeit/coadd2d.py:
```diff
@@ -1529,25 +1539,22 @@ def reference_trace_stack(self, slitid, offsets=None, objid=None):
-            containing the reference trace for each of the nexp
-            exposures.
+            ref_trace_stack (list):
+               A list of reference traces for the 2d coadding that have been offset
```
Wrong doc format.

kbwestfall replied:
Fixed.

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -400,7 +402,7 @@ def solve_poly_ratio(wave, flux, ivar, flux_ref, ivar_ref, norder, mask = None,
     tuple of outputs: tuple
            (1) ymult: `numpy.ndarray`_, (nspec,) -- rescaling factor to be
-           multiplied into flux to match flux_ref. 
+           multiplied into flux to match flux_ref.
```
wrong doc format

kbwestfall replied:
fixed.

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -1252,9 +1254,11 @@ def compute_stack(wave_grid, waves, fluxes, ivars, masks, weights, min_weight=1e
-    waves: `numpy.ndarray`_ 
+    waves: `numpy.ndarray`_
+            to a given bin will be larger.  shape=(ngrid +1,)
+    waves: `numpy.ndarray`_
```
`waves` is repeated

kbwestfall replied:
fixed.

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -1994,21 +1998,21 @@ def scale_spec_stack(wave_grid, waves, fluxes, ivars, masks, sn, weights,
         scale_spec
     scale_method_used: list
         List of methods used for rescaling spectra.
+    Returns
```
Dangling "Returns", and I think you want the empty line above between the end of the parameter list and the beginning of the returns section.

kbwestfall replied:
fixed.

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -2376,6 +2386,8 @@ def ech_combspec(waves, fluxes, ivars, masks, weights_sens, nbest=None,
            In case you want to specify the minimum wavelength in your wavelength grid, default=None computes from data.
     wave_grid_max: float, optional
            In case you want to specify the maximum wavelength in your wavelength grid, default=None computes from data.
+    wave_grid_input (`numpy.ndarray`_):
```
Wrong doc format

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -2715,9 +2726,15 @@ def get_wave_bins(thismask_stack, waveimg_stack, wave_grid):
+        wavemask = thismask & (waveimg > 1.0)
+        wave_min = waveimg[wavemask].min()
+        wave_max = waveimg[wavemask].max()
+        wave_lower = wave_min if wave_min < wave_lower else wave_lower
```
```
wave_lower = min(wave_min, wave_lower)
wave_upper = max(wave_max, wave_upper)
```

kbwestfall replied:
changed

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -2731,14 +2748,22 @@ def get_spat_bins(thismask_stack, trace_stack, spat_samp_fact=1.0):
+        are (nspec, nspat),  but each image can have a different shape.
+    ref_trace_stack: list
+        List of reference traces about which the images are rectified and coadded.  If the images were not dithered then
+        this reference trace can simply be the center of the slit:
```
Add a colon or add a `code-block`:

```
this reference trace can simply be the center of the slit::

    slitcen = (slit_left + slit_righ)/2

If the...
```

or

```
this reference trace can simply be the center of the slit:

.. code-block:: python

    slitcen = (slit_left + slit_righ)/2

If the...
```

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -2731,14 +2748,22 @@ def get_spat_bins(thismask_stack, trace_stack, spat_samp_fact=1.0):
+        or it could be the trace of the object of interest in each exposure determined by running PypeIt on the individual
+        images.  The list has nimgs elements, each of which is a 1D `numpy.ndarray`_ of shape (nspec,).
+
+    spat_samp_fact (float, optional):
```
wrong doc style

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -2753,36 +2778,30 @@ def get_spat_bins(thismask_stack, trace_stack, spat_samp_fact=1.0):
+        nspec, nspat = thismask.shape
+        spat_img = np.repeat(np.arange(nspat)[np.newaxis, :], nspec, axis=0)
+        slit_cen_img = np.repeat(trace[:,np.newaxis], nspat, axis=1)
+        dspat_iexp = (spat_img - slit_cen_img) / spat_samp_fact
```
I'm not sure about the dimensionality here, but can you do this more directly?  Something like:
```
dspat_iexp = (np.arange(nspat)[None,:] - trace[:,None]) / spat_samp_fact
```

jhennawi replied:
Implemented.

kbwestfall commented on pypeit/core/coadd.py:
```diff
@@ -2753,36 +2778,30 @@ def get_spat_bins(thismask_stack, trace_stack, spat_samp_fact=1.0):
+        spat_min_iexp = dspat_iexp[thismask].min()
+        spat_max_iexp = dspat_iexp[thismask].max()
+        spat_min = spat_min_iexp if spat_min_iexp < spat_min else spat_min
+        spat_max = spat_max_iexp if spat_max_iexp > spat_max else spat_max
```
Is there a reason you prefer the ``if ... else`` construction over using the builtin `min` and `max`?

kbwestfall replied:
Changed.

kbwestfall commented on pypeit/core/fitting.py:
```diff
@@ -407,7 +407,7 @@ def robust_fit(xarray, yarray, order, x2=None, function='polynomial',
     #pypeitFit = None
     while (not qdone) and (iIter < maxiter):
-        if np.sum(this_gpm) <= np.sum(order) + 1:
+        if (np.sum(this_gpm) <= np.sum(order) + 1):
```
You don't need the added parentheses.

kbwestfall replied:
changed

kbwestfall commented on pypeit/core/skysub.py:
```diff
@@ -48,47 +48,45 @@ def skysub_npoly(thismask):
-                  trim_edg=(3,3), pos_mask=True, show_fit=False, 
-                  no_poly=False, npoly=None):
+def global_skysub(image, ivar, tilts, thismask, slit_left, slit_righ, inmask=None, bsp=0.6, sigrej=3.0, maxiter=35,
+                  trim_edg=(3,3), pos_mask=True, max_mask_frac=0.80, show_fit=False, no_poly=False, npoly=None):
```
Please don't make lines longer.

kbwestfall replied:
changed

kbwestfall commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,219 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+        Shape must be :math:`(N_{\rm orders},)`
+    spec_arxiv :  `numpy.ndarray_`, shape=(nspec, narxiv) or (nspec)
+        Collection of archival arc spectra for which wavelength solution and line identifications are known
+    wave_soln_arxiv:  float ndarray shape (nspec, narxiv) or (nspec)
```
Does not match argument variable name.

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,219 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+def report_final(nslits, all_patt_dict, detections, wv_calib, ok_mask, bad_slits):
+    """
+    Print out the final report for wavelength calibration
+    Args:
```
You need an empty line before "Args:".

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/wavecal/autoid.py:
```diff
@@ -1127,6 +859,219 @@ def full_template(spec, lamps, par, ok_mask, det, binspectral, nsnippet=2,
+                  '  Central dispersion            = {:g}A/pix'.format(cen_disp) + msgs.newline() +
+                  '  Central wave/disp             = {:g}'.format(cen_wave / cen_disp) + msgs.newline() +
+                  '  Final RMS of fit              = {:g}'.format(wv_calib[st]['rms']))
+    return
```
Don't need the return.

kbwestfall replied:
changed

kbwestfall commented on pypeit/core/wavecal/echelle.py:
```diff
@@ -0,0 +1,224 @@
+    """
+    Predict the echelle arc spectrum using the fits to wavelength solution vs echangle and xdangle  and the archived
+    composite arcs.
+    Args:
```
Empty line above "Args:"

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/wavecal/templates.py:
```diff
@@ -391,15 +392,21 @@ def read_ascii(tbl_file, in_vac=True):
+    Parameters
+    ----------
+    xidl_file : str
+    slit : int
```
Provide brief descriptions.

jhennawi replied:
Done.

kbwestfall commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -390,63 +395,100 @@ def zerolag_shift_stretch(theta, y1, y2):
+
+    Returns:
+        xcorr_arc: numpy.ndarray_
+          Synthetic arc spectrum to be used for cross-correlations, shape = (nspec,)
```
Wrong format

kbwestfall replied:
fixed

kbwestfall commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -390,63 +395,100 @@ def zerolag_shift_stretch(theta, y1, y2):
+        sigma = fwhm/2.35
+        if tcent1[ind] == -999.0:
+            continue
+        else:
```
Don't need the else.

kbwestfall replied:
changed

kbwestfall commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -390,63 +395,100 @@ def zerolag_shift_stretch(theta, y1, y2):
 
     Args:
-        inspec1 : ndarray
+        inspec1 : numpy.ndarray_
```
You're mixing numpy and google style.  You need to stick with one.

kbwestfall replied:
fixed

kbwestfall commented on pypeit/deprecated/wvutils.py:
```diff
@@ -86,3 +87,110 @@ def hist_wavedisp(waves, disps, dispbin=None, wavebin=None, scale=1.0, debug=Fal
     return hist_wd, cent_w, cent_d
+
+def smooth_ceil_cont(inspec1, smooth, percent_ceil = None, use_raw_arc=False,sigdetect = 10.0, fwhm = 4.0):
+    """ Utility routine to smooth and apply a ceiling to spectra """
```
Needs a doc string.

jhennawi replied:
It is deprecated.

kbwestfall commented on pypeit/deprecated/wvutils.py:
```diff
@@ -86,3 +87,110 @@ def hist_wavedisp(waves, disps, dispbin=None, wavebin=None, scale=1.0, debug=Fal
+
+    Returns:
+        y1_out (ndarray):
+          Spectrum with smoothing and ceiling applied, shape = (nspec,)
```
wrong format.

jhennawi replied:
Deprecated.

kbwestfall commented on pypeit/deprecated/wvutils.py:
```diff
@@ -86,3 +87,110 @@ def hist_wavedisp(waves, disps, dispbin=None, wavebin=None, scale=1.0, debug=Fal
+    # Mask out large scale features
+    y1_ls = utils.fast_running_median(y1, fwhm * large_scale_fwhm_fact)
+    mean, med, stddev = astropy.stats.sigma_clipped_stats(
+        y1_ls[y1_ls != 0.0], sigma_lower=3.0, sigma_upper=3.0, cenfunc='median', stdfunc=utils.nan_mad_std)
```
calculate `y1_ls != 0.0` once instead of 3 times?

jhennawi replied:
It's deprecated.

kbwestfall commented on pypeit/extraction.py:
```diff
@@ -211,18 +269,20 @@ def nobj_to_extract(self):
 
-        Args:
+        Args
+            slits: :class:`SlitTraceSet` object
```
```
slits (:class:`~pypeit.slittrace.SlitTraceSet`):
```

kbwestfall replied:
fixed

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -251,18 +285,20 @@ def create_skymask(self, sobjs_obj):
         that we'll use here in :class:`FindObjects`
 
         Args:
+            slits: :class:`SlitTraceSet` object
```
```
slits (:class:`~pypeit.slittrace.SlitTraceSet`):
```

kbwestfall replied:
fixed

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -298,51 +334,63 @@ def run(self, std_trace=None, show_peaks=False):
+            return np.zeros_like(self.sciImg.image), sobjs_obj
         else:
-            msgs.info("Skipping 2nd run of finding objects")
+            # Check if the user wants to use a pre-defined sky regions file.
```
Everything after the first return doesn't need to be in an else block.

kbwestfall replied:
changed

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -298,51 +334,63 @@ def run(self, std_trace=None, show_peaks=False):
+            user_regions = self.par['reduce']['skysub']['user_regions']
+            if user_regions == '': user_regions = None
+            # Check if sky regions should be loaded
+            load_skyreg = (self.sky_region_file is not None) or (user_regions is not None)
```
don't need the parentheses

kbwestfall replied:
changed

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -298,51 +334,63 @@ def run(self, std_trace=None, show_peaks=False):
+                skymask_init = self.load_skyregions(sky_region_file=self.sky_region_file)
+                # Perform a fit to the global sky model
+                initial_sky = self.global_skysub(skymask=skymask_init, update_crmask=False, objs_not_masked=True,
+                                                  show_fit=show_skysub_fit).copy()
```
Why does this need to be copied?  `global_skysub` doesn't return an element of `self`.  Maybe it used to?

jhennawi replied:
Yes good catch. I've removed all the copies.

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -298,51 +334,63 @@ def run(self, std_trace=None, show_peaks=False):
+                # Perform a first pass sky-subtraction without masking any objects. Should  we make this no_poly=True to
+                # have fewer degrees of freedom in the with with-object global sky fits??
+                initial_sky0 = self.global_skysub( update_crmask=False, objs_not_masked=True,
+                                                  show_fit=show_skysub_fit).copy()
```
same comment about `copy`

jhennawi replied:
Fixed.

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -533,8 +584,9 @@ def global_skysub(self, skymask=None, update_crmask=True, trim_edg=(3,3),
                 no_poly=self.par['reduce']['skysub']['no_poly'],
-                pos_mask=(not self.bkg_redux) and not objs_not_masked, show_fit=show_fit)
-
+                pos_mask=(not self.bkg_redux) and not objs_not_masked,
```
don't need the parentheses

kbwestfall replied:
changed

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -298,51 +334,63 @@ def run(self, std_trace=None, show_peaks=False):
-            msgs.info("Skipping 2nd run of finding objects")
+            # Check if the user wants to use a pre-defined sky regions file.
+            user_regions = self.par['reduce']['skysub']['user_regions']
+            if user_regions == '': user_regions = None
```
Does this change the value of `self.par['reduce']['skysub']['user_regions']`?

jhennawi replied:
@rcooke-ast can you please comment? Perhaps we need a copy here since parsets should be immutable.

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -561,50 +613,60 @@ def global_skysub(self, skymask=None, update_crmask=True, trim_edg=(3,3),
-                                                      master_dir=master_dir)
+        user_regions = self.par['reduce']['skysub']['user_regions']
+        # Perform some checks
+        if user_regions is None:
```
What if `user_regions == ''` ?

jhennawi replied:
@rcooke-ast can you please comment.

rcooke-ast replied:
This possibility gets captured on line 346

kbwestfall commented on pypeit/find_objects.py:
```diff
@@ -1191,7 +1257,9 @@ def joint_skysub(self, skymask=None, update_crmask=True, trim_edg=(0,0),
                                                         bsp=self.par['reduce']['skysub']['bspline_spacing'],
                                                         no_poly=self.par['reduce']['skysub']['no_poly'],
-                                                        pos_mask=(not self.bkg_redux) and not objs_not_masked, show_fit=show_fit)
+                                                        pos_mask=(not self.bkg_redux) and not objs_not_masked,
```
don't need parentheses

kbwestfall replied:
changed

kbwestfall commented on pypeit/images/pypeitimage.py:
```diff
@@ -52,6 +53,7 @@ class PypeItImage(datamodel.DataContainer):
             Image count scaling applied (e.g., 1/flat-field)
         bpm (`numpy.ndarray`_, optional):
             Bad pixel mask
+        # TODO This docs need to indicate the types of things. Are crmask fullmask etc. boolean or ints?
```
Can you move this outside of the doc string so that it doesn't appear on readthedocs?

kbwestfall replied:
done.

kbwestfall commented on pypeit/images/pypeitimage.py:
```diff
@@ -334,16 +342,15 @@ def build_crmask(self, par, subtract_img=None):
         # If the object has multiple images, need to flag each image individually
         if self.is_multidetector:
-            self.crmask = [None] * self.shape[0]
+            self.crmask = np.empty(self.shape, dtype=bool)
```
I think there was some reason for doing it the other way, but I don't remember.

kbwestfall replied:
I'm okay with leaving these changes.

kbwestfall commented on pypeit/scripts/coadd_2dspec.py:
```diff
@@ -159,6 +159,8 @@ def main(args):
         # TODO Flexure correction for coadd2d needs to be thought through. Currently turning it off.
         parset['flexure']['spec_method'] = 'skip'
+        # TODO This is currently the default for 2d coadds, but we need a way to toggle it on/off
+        parset['reduce']['findobj']['skip_skysub'] = True
```
Do we need to warn users somehow that this is happening?  I.e., we don't want people putting something in their coadd parameter block an expect the code to respect it if we're hard-coding it this way.

jhennawi replied:
Well right now, co-adding non-skysubtracted frames is not supported, so I don't think it is a pressing issue.

kbwestfall commented on pypeit/spectrographs/jwst_nircam.py:
```diff
@@ -0,0 +1,171 @@
+        ))
+        detector_dicts = [detector_dict1, detector_dict2]
+        detector = detector_container.DetectorContainer(**detector_dicts[det-1])
+        return detector
```
Get rid of the middle man:
```
return detector_container.DetectorContainer(**detector_dicts[det-1])
```

kbwestfall replied:
done

kbwestfall commented on pypeit/spectrographs/jwst_nirspec.py:
```diff
@@ -0,0 +1,171 @@
+"""
+Module for MMT/Blue Channel specific methods.
```
Wrong spectrograph.

kbwestfall replied:
fixed

kbwestfall commented on pypeit/spectrographs/jwst_nirspec.py:
```diff
@@ -0,0 +1,171 @@
+
+class JWSTNIRSpecSpectrograph(spectrograph.Spectrograph):
+    """
+    Child to handle MMT/Blue Channel specific code
```
wrong spec

kbwestfall replied:
fixed

kbwestfall commented on pypeit/spectrographs/jwst_nirspec.py:
```diff
@@ -0,0 +1,171 @@
+
+            Many of the necessary detector parameters are read from the file
+            header, meaning the ``hdu`` argument is effectively **required** for
+            MMT/BlueChannel.  The optional use of ``hdu`` is only viable for
```
wrong spec

kbwestfall replied:
removed warning

kbwestfall commented on pypeit/spectrographs/jwst_nirspec.py:
```diff
@@ -0,0 +1,171 @@
+        ))
+        detector_dicts = [detector_dict1, detector_dict2]
+        detector = detector_container.DetectorContainer(**detector_dicts[det-1])
+        return detector
```
```
return detector_container.DetectorContainer(**detector_dicts[det-1])
```

kbwestfall replied:
changed

kbwestfall commented on pypeit/spectrographs/keck_hires.py:
```diff
@@ -94,201 +259,303 @@ def check_frame_type(self, ftype, fitstbl, exprng=None):
+            **detector_dicts[det-1])
+
+        if hdu is None:
+            return detector
```
Why do we need this?  `detector` is returned regardless of whether or not `hdu is None`.

kbwestfall replied:
changed

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -358,7 +360,7 @@ class BuildWaveCalib:
     frametype = 'wv_calib'
 
-    def __init__(self, msarc, slits, spectrograph, par, lamps, binspectral=None, det=1,
+    def __init__(self, msarc, slits, spectrograph, par, lamps, binspectral=None, meta_dict=None, det=1,
```
Add to the doc string, and describe what's in the dict that's needed by the code.

jhennawi replied:
Added.

kbwestfall commented on pypeit/wavecalib.py:
```diff
@@ -523,6 +526,26 @@ def build_wv_calib(self, arccen, method, skip_QA=False):
                                              nonlinear_counts=self.nonlinear_counts,
                                              nsnippet=self.par['nsnippet'])
                                              #debug=True, debug_reid=True, debug_xcorr=True)
+        elif self.par['method'] == 'echelle':
```
Should we instead put in a fault for now?  I.e., we don't want people using this until it's fully integrated.

jhennawi replied:
added.

kbwestfall reviewed (changes requested):
Amazing!  Lots of mostly minor comments, many of which are on the docs.

jhennawi committed:
Updating docs

kbwestfall committed:
hot fix

jhennawi committed:
Fixed a minor unit test failure.

kbwestfall committed:
Merge branch 'develop' into jwst_doc

kbwestfall committed:
doc update

kbwestfall committed:
tag docs

kbwestfall committed:
preempt change in dev-suite readme url

debora-pe committed:
Merge pull request #1492 from pypeit/staged

Merges develop into release (1.11.0 tag)

kbwestfall committed:
Merge branch 'develop' into setup_fix

kbwestfall committed:
Merge branch 'release' into setup_fix

kbwestfall committed:
CHANGES

jhennawi committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into jwst

 Conflicts:
	CHANGES.rst
	doc/api/pypeit.rst
	doc/help/run_pypeit.rst
	doc/include/datamodel_flatimages.rst
	doc/include/shane_kast_blue_A.pypeit.rst
	doc/include/spectrographs_table.rst
	doc/pypeit_par.rst
	pypeit/spectrographs/keck_hires.py
	pypeit/spectrographs/vlt_xshooter.py

kbwestfall committed:
Merge branch 'jwst' into jwst_doc

kbwestfall committed:
doc update

kbwestfall committed:
doc update

kbwestfall committed:
Merge branch 'release' into jwst_doc

kbwestfall committed:
CHANGES

kbwestfall committed:
Merge branch 'develop' into setup_bad_frames

kbwestfall committed:
Merge branch 'setup_bad_frames' into setup_fix

kbwestfall committed:
gmos wavecal note

kbwestfall committed:
generic license

jhennawi committed:
Merge branch 'setup_fix' of https://github.com/pypeit/PypeIt into jwst

kbwestfall committed:
Merge branch 'jwst' into jwst_doc

kbwestfall committed:
Merge pull request #1491 from pypeit/jwst_doc

Merged `develop` into `jwst`

kbwestfall committed:
fixed bug that assigned science frames to multiple configurations

kbwestfall committed:
easy PR fixes

kbwestfall committed:
doc fix/update

kbwestfall reviewed (commented):
Addressed all of my minor comments.

jhennawi committed:
PR comments.

jhennawi committed:
Merge branch 'jwst' of https://github.com/pypeit/PypeIt into jwst

jhennawi committed:
PR comments.

jhennawi committed:
PR comments.

bpholden committed:
Merge pull request #1499 from pypeit/setup_hotfix

(HOTFIX) Disallow science frames being assigned to multiple configurations

tbowers7 committed:
(HOTFIX) Fix typo related to GitHub download

Fixes a typo in the constructed URL of GitHub-hosted content, where a second
(spurious) "/" is included in the URL.  When AstroPy goes to download the file,
the extra slash is removed and the correct file is downloaded.  However, when
the code attempts to locate the file, it cannot find it in the cache because of
the slash mismatch.  For most users, the file is re-downloaded.  For offline
uses (for instance, compute clusters), this causes a fatal error.

	modified:   pypeit/data/utils.py

tbowers7 committed:
Add CHANGES

	modified:   CHANGES.rst

tbowers7 committed:
Merge pull request #1504 from pypeit/github_data_hotfix

(HOTFIX) Fix typo related to GitHub download

badpandabear commented:
Dev-suite passes using jwst dev-suite branch.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  201 passed, 1188 warnings in 196.01s (0:03:16) ---
--- PYTEST UNIT TESTS PASSED  118 passed, 1817 warnings in 974.17s (0:16:14) ---
--- PYTEST VET TESTS PASSED  25 passed, 1503 warnings in 840.77s (0:14:00) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 153/153 TESTS  ---
```

kbwestfall committed:
Merge branch 'release' into jwst

kbwestfall committed:
Merge branch 'develop' into jwst

kbwestfall committed:
doc fix/update
